generators/E906LegacyGen/SQPileupGen.C

```diff
@@ -0,0 +1,200 @@
+#include <E906LegacyVtxGen/SQPrimaryVertexGen.h>
+
+ExtParticle::ExtParticle(int evtID, int pdg, const TVector3& pos, const TVector3& mom):
+_evtID(evtID),
+_pdg(pdg),
+  _pos(pos),
+  _mom(mom)
+{
+}
+
+SQPileupGen::SQPileupGen(const std::string& name):
+  PHG4ParticleGeneratorBase(name),
+  _ineve(nullptr),
+  _mcevt(nullptr),
+  _extFile(nullptr),
+  _extTree(nullptr),
+  _extPos(nullptr),
+  _extMom(nullptr),
+  _bucketSize(0),
+  _evt(0),
+  _beam_intensity_profile(nullptr),
+  _beam_intensity_profile_histo(nullptr),
+  _n_proc_evt(0),
+  _inhibit_threshold(5000),
+  _proton_coeff(0.03),
+  _vertexGen(new SQPrimaryVertexGen)
+{
+}
+
+SQPileupGen::~SQPileupGen()
+{
+  if(_vertexGen != nullptr) delete _vertexGen;
+}
+
+int SQPileupGen::Init(PHCompositeNode* topNode)
+{  
+  return Fun4AllReturnCodes::EVENT_OK;
+}
+
+int SQPileupGen::InitRun(PHCompositeNode* topNode)
+{
+
+  gRandom->SetSeed(PHRandomSeed());
+  _ineve = findNode::getClass<PHG4InEvent>(topNode, "PHG4INEVENT");
+  if(!_ineve) 
+    {
+      PHNodeIterator iter(topNode);
+      PHCompositeNode* dstNode = dynamic_cast<PHCompositeNode*>(iter.findFirst("PHCompositeNode", "DST"));
+      
+      _ineve = new PHG4InEvent();
+      dstNode->addNode(new PHDataNode<PHObject>(_ineve, "PHG4INEVENT", "PHObject"));
+    }
+
+  _evt = findNode::getClass<SQEvent>(topNode, "SQEvent");
+  if (! _evt) {
+    PHNodeIterator iter(topNode);
+    PHCompositeNode* dstNode = dynamic_cast<PHCompositeNode*>(iter.findFirst("PHCompositeNode", "DST"));
+
+    _evt = new SQEvent_v1();
+    dstNode->addNode(new PHIODataNode<PHObject>(_evt, "SQEvent", "PHObject"));
+  }
+
+  _mcevt = findNode::getClass<SQMCEvent>(topNode, "SQMCEvent");
+  if(!_mcevt) 
+    {
+      PHNodeIterator iter(topNode);
+      PHCompositeNode* dstNode = dynamic_cast<PHCompositeNode*>(iter.findFirst("PHCompositeNode", "DST"));
+
+      _mcevt = new SQMCEvent_v1();
+      dstNode->addNode(new PHIODataNode<PHObject>(_mcevt, "SQMCEvent", "PHObject"));
+    }
+
+  _extFile = TFile::Open(_extFileName.Data());
+  if(!_extFile)
+    {
+      std::cout << "SQPileupGen: cannot find the input ROOT file, abort " << _extFileName << std::endl;
+      exit(EXIT_FAILURE);
+    }
+  _extTree = (TTree*)_extFile->Get("bkg");
+
+  _extPos = new TClonesArray("TVector3");
+  _extMom = new TClonesArray("TVector3");
+
+  _extTree->SetBranchAddress("eventID", &_extEvtID);
+  _extTree->SetBranchAddress("n", &_nExtPar);
+  _extTree->SetBranchAddress("pdg", _extPDG);
+  _extTree->SetBranchAddress("pos", &_extPos);
+  _extTree->SetBranchAddress("mom", &_extMom);
+
+  _readIdx = 0;
+
+  _vertexGen->InitRun(topNode);
+  return Fun4AllReturnCodes::EVENT_OK;
+}
+
+int SQPileupGen::process_event(PHCompositeNode* topNode)
+{
+  double QIEcount;
+  
+  if(_beam_intensity_profile) 
+    {
+      do {
+       QIEcount = _beam_intensity_profile->GetRandom();
+      } while (QIEcount>_inhibit_threshold);
+      _bucketSize  = (int)QIEcount/_proton_coeff;
```

Comment:
Dear Kenichi,
Yes I thought you wanted me to delete the lines, because we were repeating same setBucketSize for the function and histogram case. I realized that we can simply set the _bucketSize value to take care of both case. Also, if we put that outside the loop then we will mess up while the case when set bucketsize by a constant number.

Reply:
I see.  I first wanted to put the lines outside the loop, and I didn't take into account the constant number case.  You are correct and the present code is fine.  It is helpful if you could explain your thought when making such change in the future.